crates/wasmi/src/memory/buffer.rs

```diff
@@ -20,33 +27,157 @@ impl ByteBuffer {
     /// - If the initial length is 0.
     /// - If the initial length exceeds the maximum supported limit.
     pub fn new(initial_len: usize) -> Self {
+        let mut vec = vec![0x00_u8; initial_len];
+        let (ptr, len, capacity) = (vec.as_mut_ptr(), vec.len(), vec.capacity());
+        core::mem::forget(vec);
         Self {
-            bytes: vec![0x00_u8; initial_len],
+            ptr,
+            len,
+            capacity,
+            is_static: false,
+        }
+    }
+
+    /// Creates a new byte buffer with the given initial length.
+    ///
+    /// # Errors
+    ///
+    /// - If the initial length is 0.
+    /// - If the initial length exceeds the maximum supported limit.
+    pub fn new_static(buf: &'static mut [u8], initial_len: usize) -> Self {
+        assert!(initial_len <= buf.len());
+        Self {
+            ptr: buf.as_mut_ptr(),
+            len: initial_len,
+            capacity: buf.len(),
+            is_static: true,
         }
     }
 
     /// Grows the byte buffer to the given `new_size`.
     ///
     /// # Panics
     ///
-    /// If the current size of the [`ByteBuffer`] is larger than `new_size`.
+    /// - If the current size of the [`ByteBuffer`] is larger than `new_size`.
+    /// - If backed by static buffer and `new_size` is larger than it's capacity.
     pub fn grow(&mut self, new_size: usize) {
         assert!(new_size >= self.len());
-        self.bytes.resize(new_size, 0x00_u8);
+        if self.is_static {
+            if self.capacity < new_size {
+                panic!("Cannot grow static byte buffer more then it's capacity")
+            }
+            // Safety:
+            // - self.ptr.add(self.len) is still in bounds of the buffer.
+            // - we are allowed to write to the buffer anywhere between `len` and `capacity`.
+            unsafe {
+                core::ptr::write_bytes(self.ptr.add(self.len), 0x00_u8, new_size - self.len);
+            }
+            self.len = new_size;
```

Comment:
```suggestion
            let old_size = self.len;
            self.len = new_size;
            let data = self.data_mut()[old_size..].fill(0x0);
```

Reply:
I have not checked if the above code actually works. Please check yourself when you apply the changes. :)